app/scripts/modules/core/src/instance/Instances.tsx

```diff
@@ -1,6 +1,133 @@
+  // context from enclosing UIView
+  public static contextTypes = {
+    router: React.PropTypes.object,
+    parentUIViewAddress: React.PropTypes.object,
+  };
+
+  private subscription: Subscription;
+
+  constructor(props: IInstancesProps) {
+    super(props);
+    this.state = {
+      detailsInstanceId: null,
+    }
+  }
+
+  public render() {
+    const instances = (this.props.instances || []).sort((a, b) => a.launchTime - b.launchTime);
+
+    return (
+      <div className="instances">
+        {instances.map(instance => (
+          <Instance
+            key={instance.id}
+            instance={instance}
+            active={this.state.detailsInstanceId === instance.id}
+            highlight={this.props.highlight}
+            onInstanceClicked={this.handleInstanceClicked}
+          />
+        ))}
+      </div>
+    )
+  }
+
+  public componentDidMount() {
+    this.subscription = this.context.router.globals.success$.subscribe(this.onStateChange);
+  }
+
+  private onStateChange(transition: Transition) {
+    const isShowingDetails = !!/\.instanceDetails/.exec(transition.to().name);
+    const detailsInstanceId = isShowingDetails ? transition.params().instanceId : null;
+    this.setState({ detailsInstanceId });
+  }
+
+  public componentWillUnmount() {
+    this.subscription && this.subscription.unsubscribe();
+  }
+
+  public shouldComponentUpdate(nextProps: IInstancesProps, nextState: IInstancesState) {
+    const propsKeys: [keyof IInstancesProps] = ['instances', 'highlight'];
+    if (!equalForKeys(propsKeys, this.props, nextProps)) {
+      return true;
+    }
+
+    if (this.state.detailsInstanceId !== nextState.detailsInstanceId) {
+      const ids = nextProps.instances.map(x => x.id);
+      if (ids.includes(this.state.detailsInstanceId) || ids.includes(nextState.detailsInstanceId)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private handleInstanceClicked(instance: IInstance) {
+    const { router, parentUIViewAddress } = this.context;
+    const params = { instanceId: instance.id, provider: instance.provider };
+    const options = { relative: parentUIViewAddress.context };
+    router.stateService.go('.instanceDetails', params, options);
+  }
+}
+
+export interface IInstanceProps {
+  instance: IInstance;
+  active: boolean;
+  highlight: string;
+  onInstanceClicked(instance: IInstance): void;
+}
+
+@autoBindMethods
+export class Instance extends React.Component<IInstanceProps, void> {
+  public render() {
+    const { instance, active, highlight } = this.props;
+    const { id, healthState } = instance;
+
+    const className = `instance health-status-${healthState} ${highlight === id ? 'highlighted' : ''} ${active ? 'active' : ''}`;
+    return (
+      <a
+        className={className}
+        title={id}
+        data-toggle="tooltip"
+        onMouseOver={this.onMouseOver}
+        onClick={this.handleClick}
```

Comment:
Curious to see how much of the computation time is spent attaching these handlers to each instance...

I'm going to run this PR locally and see if delegating mouse events to the parent container makes much difference...